### PR TITLE
feat(gltf): add name component to gltf mesh primitive

### PR DESCRIPTION
**Pull request by Soulghost**
# Objective

- fixes https://github.com/bevyengine/bevy/issues/13473

## Solution

- When a single mesh is assigned multiple materials, it is divided into several primitive nodes, with each primitive assigned a unique material. Presently, these primitives are named using the format Mesh.index, which complicates querying. To improve this, we can assign a specific name to each primitive based on the material’s name, since each primitive corresponds to one material exclusively.

## Testing

- I have included a simple example which shows how to query a material and mesh part based on the new name component.

## Changelog
- adds `GltfMaterialName` component to the mesh entity of the gltf primitive node.

### CODE REVIEW

File: examples/3d/query_gltf_primitives.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            num_cascades: 1,
+            maximum_distance: 1.6,
+            ..default()
+        }
+        .into(),
+        ..default()
+    });
+    commands.spawn(SceneBundle {
+        scene: asset_server.load(
+            GltfAssetLabel::Scene(0).from_asset("models/GltfPrimitives/gltf_primitives.glb#Scene0"),
```
Comment (alice-i-cecile) [inline examples/3d/query_gltf_primitives.rs:53]:
```suggestion
            GltfAssetLabel::Scene(0).from_asset("models/GltfPrimitives/gltf_primitives.glb"),
```
That should be redundant.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Useful, but I this should just use the standard Bevy `Name` component. That reduces the code that we have to maintain, and will play nicer with various tooling like `bevy_inspector_egui`.

I'm also having a hard time figuring out how the example demonstrates the added functionality. Can you explain what you were after? I may be able to suggest how to make things clearer.

**Comment by Soulghost** [comment]
> Useful, but I this should just use the standard Bevy `Name` component. That reduces the code that we have to maintain, and will play nicer with various tooling like `bevy_inspector_egui`.
> 
> I'm also having a hard time figuring out how the example demonstrates the added functionality. Can you explain what you were after? I may be able to suggest how to make things clearer.

Hi Alice, I apologize for the confusion caused by the buggy example `query_gltf_primitives`. As detailed in the issue https://github.com/bevyengine/bevy/issues/13473, the newly introduced `GltfMaterialName` component is used to query for a mesh part with the specific material. When multiple materials are assigned to different parts of a mesh, the corresponding primitives are generated in equal number. However, these primitives currently lack individual names.

```json
{
  "meshes": [
    {
      "name": "Cube",
      "primitives": [
        {
          "attributes": { "POSITION": 0, "NORMAL": 1, "TEXCOORD_0": 2 },
          "indices": 3,
          "material": 0
        },
        {
          "attributes": { "POSITION": 4, "NORMAL": 5, "TEXCOORD_0": 6 },
          "indices": 7,
          "material": 1
        },
        {
          "attributes": { "POSITION": 8, "NORMAL": 9, "TEXCOORD_0": 10 },
          "indices": 11,
          "material": 2
        },
        {
          "attributes": { "POSITION": 12, "NORMAL": 13, "TEXCOORD_0": 14 },
          "indices": 15,
          "material": 3
        }
      ]
    }
  ]
}

```

In bevy, each primitive is named according to its index.

```rust
// crates/bevy_gltf/src/loader.rs
fn primitive_name(mesh: &gltf::Mesh, primitive: &Primitive) -> String {
    let mesh_name = mesh.name().unwrap_or("Mesh");
    if mesh.primitives().len() > 1 {
        format!("{}.{}", mesh_name, primitive.index())
    } else {
        mesh_name.to_string()
    }
}
```

This causes a problem, which is addressed by @MatrixDev in the above issue:
> I don't know which primitive contains which material. In case I want to motify some named material, I have to play with scene directly, used entities mapping, it's a boilerplate mess and not ECS friendly.

So it appears useful to name each primitive according to its associated material, given that primitives are distinguished based on the number of materials assigned to a single mesh. However, since some may rely on the indexed names stored in the `Name` component, I decided to introduce a new naming component for this purpose. This may not be the optimal solution, and I would appreciate your suggestions. Thank you.

**Comment by MatrixDev** [comment]
I think using 'Name' for everything is not enough. Entity can have a GLTF node/primitive name/index and material name/index all at the same time.

We should be able to query using any of those at any time. Sometimes I want to query for all entities with a specific material, other times I'd like to query for all entities with a specific primitive. Names of materials and primitives might be different.

I'm not even talking about bevy generating names for entites when nodes contain multiple primitives.

**Comment by alice-i-cecile** [comment]
Right 🤔 Okay, I see your point. Thank you for explaining!

File: examples/3d/query_gltf_primitives.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
+//! Query primitives in a glTF scene.
```
Comment (alice-i-cecile) [inline examples/3d/query_gltf_primitives.rs:1]: This description needs work.

File: examples/3d/query_gltf_primitives.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            .load(GltfAssetLabel::Scene(0).from_asset("models/GltfPrimitives/gltf_primitives.glb")),
+        transform: Transform {
+            rotation: Quat::from_rotation_y(-90.0 / 180.0 * std::f32::consts::PI),
+            ..default()
+        },
+        ..default()
+    });
+}
+
+fn random_color_and_position(
```
Comment (alice-i-cecile) [inline examples/3d/query_gltf_primitives.rs:62]: This system should be moved up and renamed to make the point more clear. Something like "find_top_material" would be much more helpful.

File: crates/bevy_gltf/src/lib.rs [hunk 4]
```diff
@@ -378,6 +379,11 @@ pub struct GltfMaterialExtras {
     pub value: String,
 }
 
+/// The material name of a gltf primitive.
```
Comment (alice-i-cecile) [inline crates/bevy_gltf/src/lib.rs:382]: This should link to the glTF's spec's section on materials, and use the appropriate capitalization.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I think I'm on board now, but there's some work to be done to make this clearer and more beginner friendly.

**Comment by Soulghost** [comment]
I have improved the documentation, corrected capitalization errors, and optimized the code of the `find_top_material_and_mesh` system to achieve smooth color transitions and vertex transformations.

**Comment by alice-i-cecile** [comment]
Fantastic, thank you!

**Review by BenjaminBrienen — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1703 if you'd like to help out.

**Pull request merged**